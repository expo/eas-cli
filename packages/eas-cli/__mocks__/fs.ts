import { fs } from 'memfs';

// needed because of a weird bug with tempy (dependency of @expo/config-plugins)
// eslint-disable-next-line node/no-sync
fs.mkdirSync('/tmp');
if (process.env.TMPDIR) {
  // eslint-disable-next-line node/no-sync
  fs.mkdirSync(process.env.TMPDIR, { recursive: true });
}

// fs-extra@10 is not compatible with memfs, the following line fixes tests
(fs.realpath as any).native = jest.fn();

module.exports = fs;
