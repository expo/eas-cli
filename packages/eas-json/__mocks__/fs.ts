import { fs } from 'memfs';

// fs-extra@10 is not compatible with memfs, the following line fixes tests
(fs.realpath as any).native = jest.fn();

module.exports = fs;
