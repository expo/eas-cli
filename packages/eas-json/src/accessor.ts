import { codeFrameColumns } from '@babel/code-frame';
import assert from 'assert';
import chalk from 'chalk';
import fs from 'fs-extra';
import * as fleece from 'golden-fleece';
import path from 'path';

import { InvalidEasJsonError, MissingEasJsonError } from './errors';
import { EasJsonSchema } from './schema';
import { EasJson } from './types';

export class EasJsonAccessor {
  private easJsonPath: string;

  private isJson5 = false;
  private easJson: EasJson | undefined;
  private easJsonRawContents: string | undefined;
  private easJsonRawObject: any | undefined;
  private easJsonPatched: boolean = false;

  constructor(projectDir: string) {
    this.easJsonPath = EasJsonAccessor.formatEasJsonPath(projectDir);
  }

  public static formatEasJsonPath(projectDir: string): string {
    return path.join(projectDir, 'eas.json');
  }

  public async readAsync(): Promise<EasJson> {
    if (this.easJson) {
      return this.easJson;
    }

    const rawJSON = await this.readRawJsonAsync();
    const { value, error } = EasJsonSchema.validate(rawJSON, {
      allowUnknown: false,
      abortEarly: false,
      convert: true,
      noDefaults: true,
    });
    if (error) {
      const errorMessages = error.message.split('. ');
      throw new InvalidEasJsonError(
        `${chalk.bold('eas.json')} is not valid.\n- ${errorMessages.join('\n- ')}`
      );
    }
    this.easJson = value;
    return value;
  }

  public async writeAsync(): Promise<void> {
    if (!this.easJsonPatched) {
      return;
    }
    await fs.writeFile(this.easJsonPath, this.easJsonRawContents);
    this.resetState();
  }

  public patch(fn: (easJsonRawObject: any) => any): void {
    assert(
      this.easJsonRawContents && this.easJsonRawObject,
      'call readAsync/readRawJsonAsync first'
    );

    this.easJsonRawObject = fn(this.easJsonRawObject);
    if (this.isJson5) {
      this.easJsonRawContents = fleece.patch(this.easJsonRawContents, this.easJsonRawObject);
    } else {
      this.easJsonRawContents = `${JSON.stringify(this.easJsonRawObject, null, 2)}\n`;
    }
    this.easJsonPatched = true;
  }

  public async readRawJsonAsync(): Promise<any> {
    if (!(await fs.pathExists(this.easJsonPath))) {
      throw new MissingEasJsonError(
        `${chalk.bold('eas.json')} could not be found at ${
          this.easJsonPath
        }. Learn more at https://expo.fyi/eas-json`
      );
    }

    this.easJsonRawContents = await fs.readFile(this.easJsonPath, 'utf-8');

    try {
      const rawJSON = JSON.parse(this.easJsonRawContents);
      this.easJsonRawObject = rawJSON;
      return rawJSON;
    } catch {
      // ignore error, try reading as JSON5
    }

    try {
      const rawJSON = fleece.evaluate(this.easJsonRawContents);
      this.easJsonRawObject = rawJSON;
      this.isJson5 = true;
      return rawJSON;
    } catch (originalError: any) {
      const err = new InvalidEasJsonError(`Found invalid character in ${chalk.bold('eas.json')}.`);
      if (originalError.loc) {
        const codeFrame = codeFrameColumns(this.easJsonRawContents, { start: originalError.loc });
        err.message += `\n${codeFrame}`;
      }
      throw err;
    }
  }

  private resetState(): void {
    this.isJson5 = false;
    this.easJson = undefined;
    this.easJsonRawContents = undefined;
    this.easJsonRawObject = undefined;
    this.easJsonPatched = false;
  }
}
