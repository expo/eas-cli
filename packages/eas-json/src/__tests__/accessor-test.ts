import fs from 'fs-extra';
import { vol } from 'memfs';
import path from 'path';

import { EasJsonAccessor } from '../accessor';

const fixturesDir = path.join(__dirname, 'fixtures');
const fakeAppPath = '/fake/path/app';

const fsReal = jest.requireActual('fs').promises as typeof fs;
jest.mock('fs');

beforeEach(async () => {
  vol.reset();
});

describe(EasJsonAccessor, () => {
  test('patching JSON file', async () => {
    vol.fromJSON(
      {
        './eas.json': await fsReal.readFile(path.join(fixturesDir, 'eas-json.json'), 'utf-8'),
      },
      fakeAppPath
    );

    const accessor = new EasJsonAccessor(fakeAppPath);
    await accessor.readAsync();
    accessor.patch(o => {
      o.build.production.env.ABC = '456';
      return o;
    });
    await accessor.writeAsync();

    const newEasJsonContents = await fs.readFile(path.join(fakeAppPath, 'eas.json'), 'utf-8');
    expect(newEasJsonContents).toMatchSnapshot();
  });

  test('patching JSON5 file (preserves comments)', async () => {
    vol.fromJSON(
      {
        './eas.json': await fsReal.readFile(path.join(fixturesDir, 'eas-json5.json'), 'utf-8'),
      },
      fakeAppPath
    );

    const accessor = new EasJsonAccessor(fakeAppPath);
    await accessor.readAsync();
    accessor.patch(o => {
      o.build.production.env.ABC = '456';
      return o;
    });
    await accessor.writeAsync();

    const newEasJsonContents = await fs.readFile(path.join(fakeAppPath, 'eas.json'), 'utf-8');
    expect(newEasJsonContents).toMatchSnapshot();
  });

  test('reading invalid JSON5 object', async () => {
    vol.fromJSON(
      {
        './eas.json': await fsReal.readFile(
          path.join(fixturesDir, 'eas-invalid-json5.json'),
          'utf-8'
        ),
      },
      fakeAppPath
    );

    const accessor = new EasJsonAccessor(fakeAppPath);
    await expect(accessor.readAsync()).rejects.toThrow(/Found invalid JSON in eas.json/);
  });
});
