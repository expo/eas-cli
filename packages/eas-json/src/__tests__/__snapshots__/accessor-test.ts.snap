// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EasJsonAccessor patching JSON file 1`] = `
"{
  \\"build\\": {
    \\"production\\": {
      \\"env\\": {
        \\"TEST\\": \\"123\\",
        \\"ABC\\": \\"456\\"
      }
    }
  }
}
"
`;

exports[`EasJsonAccessor patching JSON5 file (preserves comments) 1`] = `
"{
  \\"build\\": {
    \\"production\\": {
      // comment
      \\"env\\": {
        \\"TEST\\": \\"123\\",
        ABC: \\"456\\"
      }
    }
  }
}
"
`;
